lib/oli_web/live/delivery/instructor_dashboard/instructor_dashboard_live.ex

```diff
@@ -0,0 +1,146 @@
+defmodule OliWeb.Delivery.InstructorDashboard.InstructorDashboardLive do
+  use OliWeb, :live_view
+  alias OliWeb.Components.Delivery.InstructorDashboard
+  alias OliWeb.Components.Delivery.CourseContentPanel
+  alias alias Oli.Delivery.Sections
+  alias Oli.Publishing.DeliveryResolver
+  alias Oli.Resources.Collaboration
+  use OliWeb.Common.Modal
+
+  @impl Phoenix.LiveView
+  def mount(_params, _session, socket) do
+    {:ok, socket}
+  end
+
+  @impl Phoenix.LiveView
+  def handle_params(params, _, socket) do
+    {:noreply,
+     assign(socket, params: params, active_tab: String.to_existing_atom(params["active_tab"]))}
+  end
+
+  @impl Phoenix.LiveView
+  def render(assigns) do
+    ~H"""
+    <InstructorDashboard.main_layout {assigns}>
+      <InstructorDashboard.tabs active_tab={@active_tab} section_slug={@section_slug} preview_mode={@preview_mode} />
+      <%= render_tab(assigns) %>
+    </InstructorDashboard.main_layout>
+    """
+  end
+
+  defp render_tab(%{active_tab: :students} = assigns) do
+    ~H"""
+      <.live_component
+      id="students_table"
+      module={OliWeb.Components.Delivery.Students}
+      params={@params}
+      context={@context}
+      section={@section}
+      />
+    """
+  end
+
+  defp render_tab(%{active_tab: :content} = assigns) do
+    page_link_url =
+      if assigns.preview_mode do
+        &Routes.page_delivery_path(
+          OliWeb.Endpoint,
+          :page_preview,
+          assigns.section_slug,
+          &1
+        )
+      else
+        &Routes.page_delivery_path(OliWeb.Endpoint, :page, assigns.section_slug, &1)
+      end
+
+    container_link_url =
+      if assigns.preview_mode do
+        &Routes.page_delivery_path(
+          OliWeb.Endpoint,
+          :container_preview,
+          assigns.section_slug,
+          &1
+        )
+      else
+        &Routes.page_delivery_path(OliWeb.Endpoint, :container, assigns.section_slug, &1)
+      end
+
+    assigns =
+      Map.merge(
+        assigns,
+        %{
+          hierarchy: Sections.build_hierarchy(assigns.section),
```

Comment:
This seems suspicious: Within the rendering step we are calling a fairly expensive `build_hierarchy/1`.  I see other places in this LV where DB calls are made within the rendering code.  Unless we have strong guarantees here, this code could get executed way more than what we would like.  

And we probably do not even want this DB code to execute when tabs are switched, correct?  Instead, I think what we want is a state holder within this LV that caches the state needed for each tab (live component), and that only executes the DB code that a tab needs *one* time, the very first time that a tab is displayed.

Reply:
Let me describe a scenario and you can let me know if what you have built here supports this or not:

Imagine this LV approach paired with four tabs.  Each tab has a corresponding live_component that displays a searchable, filterable table.  Each component tab must run an expensive query to fetch all the data that it needs to display.  When a user firsts visits this LiveView, there will of course be a default tab that is displayed.   Then, after the user clicks to select another tab, the query related to that other tab is run.  When the user clicks back to the original, default tab its query is **not re-run.**  The original data from its initial execution is preserved and displayed.  

What we really need also is to support two different types of live_component tab implementations:

The first implementation type is where the DB queries required to fetch the data to display in the table is so expensive that paging, filtering and sorting must be done across the entire dataset, in memory.  An example of this would be the "content" tab.  Another requirement is that the number of total records should be reasonably small (for instance, we know that a course will have at most around 50 or 60 modules, so the "content" tab qualifies against this criteria).

The other type of live_component tab provider is one where the query to fetch data is not as expensive, and perhaps the number of records can be so large that we must do the paging, sorting and filtering at the DB query level.  

For the first implementation type, we want a strong guarantee that the query to fetch all the data is run exactly one time for the entire lifetime of the parent LiveView.  So that means the first time that a user clicks on one of these types of tab the query is run then.  But afterwards, it is never run again.  The user can sort, filter, page and all that is being done in memory.  The user can click away to another tab, then click back and the query is not re-run.  The original full data results are cached in the parent LiveView.